internal/controller/controllers/bmh_agent_controller.go

```diff
@@ -384,6 +392,91 @@ func (r *BMACReconciler) reconcileBMH(ctx context.Context, bmh *bmh_v1alpha1.Bar
 	return reconcileComplete{}
 }
 
+// Finds the installation disk based on the RootDeviceHints
+//
+// This function implements the logic to find the installation disk following what's currently
+// supported by OpenShift, instead of *all* the supported cases in Ironic. The following link
+// points to the RootDeviceDisk translation done by the BareMetal Operator that is then sent to
+// Ironic:
+// https://github.com/metal3-io/baremetal-operator/blob/dbe8780ad14f53132ba606d1baec808997febe49/pkg/provisioner/ironic/devicehints/devicehints.go#L11-L54
+//
+// The logic is quite straightforward and the checks done match what is in the aforementioned link.
+// Some string checks require equality, others partial equality, whereas the int checks require numeric comparison.
+//
+// Ironic's internal filter process requires that all the disks have to fully match the RootDeviceHints (and operation),
+// which is what this function does.
+//
+// This function also filters out disks that are not elegible for installation, as we already know those cannot be used.
+func (r *BMACReconciler) findInstallationDiskID(devices []adiiov1alpha1.HostDisk, hints *bmh_v1alpha1.RootDeviceHints) string {
+	if hints == nil {
+		return ""
+	}
+
+	for _, disk := range devices {
+
+		if !disk.InstallationEligibility.Eligible {
+			continue
+		}
+
+		if hints.DeviceName != "" && hints.DeviceName != disk.Path {
+			continue
+		}
+
+		if hints.HCTL != "" && hints.HCTL != disk.Hctl {
+			continue
+		}
+
+		if hints.Model != "" && !strings.Contains(disk.Model, hints.Model) {
+			continue
+		}
+
+		if hints.Vendor != "" && !strings.Contains(disk.Vendor, hints.Model) {
+			continue
+		}
+
+		if hints.SerialNumber != "" && hints.SerialNumber != disk.Serial {
+			continue
+		}
+
+		if hints.MinSizeGigabytes != 0 {
+			sizeGB := int(disk.SizeBytes / (1024 * 3))
+			if hints.MinSizeGigabytes < sizeGB {
```

Comment:
This logic seems reversed to me.. Or am I just too sleepy?

Reply:
Nope, this is definitely backwards.

Reply:
https://bugzilla.redhat.com/show_bug.cgi?id=2131017
https://github.com/openshift/assisted-service/pull/4452